crossvalidation.py

```diff
@@ -2,123 +2,188 @@
+        description="""regularization tuning  of Product Space FEM 
+            using cross validation
+        """
     )
-    train_loss = pf.LossFunctional(train_sd, control, params['regularization'])
-    
-    xy0, xy1 = test.pair_xy()
-    test_sd = pf.SpatialData(
-            test.genetic_data["divergence"].to_numpy(),
-            xy0, xy1,
-            W,
+
+    parser.add_argument(
+        "--json",
+        type=str,
+        required=True,
+        help="""Path to JSON file with.
+        should include:
+            "spatial_data", "genetic_data" : file paths
+            "mesh" : either an XML file or a [width, height] of integers
+            "folds": number of folds
+            "regularization": {{ "l2": [a, b], "smoothing": [c, d] }}
+                where the second value for each is the *ratio* of regularization
+                strengths for the two controls (ellipse and vector)
+            "boundary": {{ "epsilon": x, "eps0": y, "eps1": z }}
+            "method": method for optimizing (default: BFGS)
+            "options": additional options to the optimizer
+        """,
+    )
+
+    parser.add_argument(
+        "-H",
+        "--use_hyperopt",
+        action="store_true",
+        help="""Switch to tune parameters using hyperopt instead of grid search.
+    """,
+    )
+
+    parser.add_argument(
+        "-e",
+        "--max_evals",
+        type=int,
+        default=100,
+        help="""Maximum number of evaluations for hyperopt.""",
+    )
+
+    parser.add_argument(
+        "-i",
+        "--max_iter",
+        type=int,
+        default=100,
+        help="""Maximum number of iterations for model optimization.""",
     )
-    test_loss = pf.LossFunctional(test_sd, control, params['regularization'])
+
+    return parser.parse_args()
+
+
+def objective(tuning_params = None):
+    # defaults
+    params = {
+        "method": "BFGS",
+        "options": {
+            'gtol': 1e-8,
+            'xrtol': 1e-8,
+            'maxiter': args.max_iter,
+        },
+        "boundary": None,
+    }
+
+    paramsfile = args.json
+    outdir = os.path.dirname(paramsfile)
+    with open(paramsfile, 'r') as f:
+        file_params = json.load(f)
+
+    params.update(file_params)
     
-    invp = pf.InverseProblem(eqn, train_loss)
-    m_hats, losses, optim_return = invp.optimize(
-            control, 
-            method=params['method'],
-            options=params['options'],
+    if args.use_hyperopt:
+        results_file = os.path.join(outdir, 'results_hyperopt.pkl')
+    else:
+        results_file = os.path.join(outdir, 'results.pkl')
+
+    # load spatial and genetic data
+    spatial_data = pd.read_csv(os.path.join(outdir, params['spatial_data'])).rename(
+            columns={"site_name": "name", "long": "x", "lat": "y"}
+    )
+    genetic_data = pd.read_csv(os.path.join(outdir, params['genetic_data'])).rename(
+            columns={"loc1": "name1", "loc2": "name2", "dxy": "divergence"}
```

Comment:
shouldn't this be outside the `objective` function? As written, won't it re-load the CSV files every iteration of `fmin`?